# ID Ryan (Ran) Wright 305109613
# ID Ben Remez 315470450
# LINK to Github Repo: (https://github.com/RyanWri/afeka-nlp-homeworks)
## repo is private, request access if needed

------------------------------------------------------------------------------------------
# Section 1

In [14]:
from word_2_vec import read_sentences

file_path = "data/sentences.txt"
sentences = read_sentences(file_path)

In [15]:
# 1a
from word_2_vec import train_skipgram_model

skip_gram_model = train_skipgram_model(
    sentences, vector_size=128, window=3, min_count=0
)

In [16]:
# 1b
from word_2_vec import train_cbow_model

cbow_model = train_cbow_model(sentences, vector_size=128, window=3, min_count=0)

In [17]:
# 1c
# Save the models for later use
skip_gram_model.save("models/skipgram_model_sentences.model")
cbow_model.save("models/cbow_model_sentences.model")

In [18]:
# 1d What is the difference between the embeddings? explain the results.
from word_2_vec import save_embeddings

# Save embeddings for Skip-gram model
skipgram_words, skipgram_embeddings = save_embeddings(
    skip_gram_model, "embeddings/skipgram_embeddings.npy"
)

# Save embeddings for CBOW model
cbow_words, cbow_embeddings = save_embeddings(
    cbow_model, "embeddings/cbow_embeddings.npy"
)

# Example: Print the embeddings for the word 'bank' from both models
print("Skip-gram embedding for 'bank':", skip_gram_model.wv["bank"])
print("CBOW embedding for 'bank':", cbow_model.wv["bank"])

Skip-gram embedding for 'bank': [ 7.5683417e-03 -7.2578448e-03  7.1992376e-03 -7.2501525e-03
 -5.3959396e-03 -7.1142283e-03 -4.3308940e-03  5.7557807e-03
  7.1583716e-03 -2.5981653e-03  2.9124138e-03 -2.8304937e-03
  6.1557195e-03  4.5791655e-03 -1.2418577e-06 -2.8341634e-03
 -5.6440174e-03  3.7270307e-03  1.1389642e-03 -2.0413594e-03
  6.1251894e-03 -3.1593211e-03 -7.1500568e-03 -1.7657378e-03
  9.6169206e-05 -5.1883911e-03 -4.2913891e-03 -6.6394489e-03
  7.2110756e-03  5.8018365e-03 -2.3181770e-04  5.7576098e-03
  6.2090098e-03 -6.0985476e-04  5.1686978e-03  2.9448657e-03
  3.9690570e-03  5.6682522e-03 -3.6989313e-03 -1.7068783e-03
  6.8067422e-04  3.3117849e-03  2.5822087e-03  3.9831204e-03
  3.5844159e-03 -6.5902080e-03 -2.4857672e-03 -5.6565059e-03
  7.5593605e-03  3.9111595e-03  1.3734704e-04  3.2150068e-03
 -5.9818849e-03 -4.9173953e-03  2.4072248e-03  5.1054368e-03
  3.0876258e-03  4.6997718e-03 -1.5533476e-03 -2.6146772e-03
  1.6406385e-04 -2.5013285e-03 -4.3083667e-03 -6.0840

***1e*** Can you find a text that its embedding will be similar in these two algorithms?<br/>
sentence -> The cat sat on the mat.<br/>
explanation -> To summarize, texts with consistent word usage and simple sentence structures, <br/>
with little to no ambiguity in word meaning, are likely to produce similar embeddings 
in both Skip-gram and CBOW models. 

In [19]:
# 1f Repeat step c with different window sizes. Is there a significant change?
window = 5
skip_gram_model2 = train_skipgram_model(
    sentences, vector_size=128, window=window, min_count=0
)
skip_gram_model2.save("models/skipgram_model_sentences_2.model")
skipgram_words_2, skipgram_embeddings_2 = save_embeddings(
    skip_gram_model2, "embeddings/skipgram_embeddings_2.npy"
)

cbow_model2 = train_cbow_model(sentences, vector_size=128, window=window, min_count=0)
cbow_model2.save("models/cbow_model_sentences_2.model")
cbow_words2, cbow_embeddings2 = save_embeddings(
    cbow_model2, "embeddings/cbow_embeddings_2.npy"
)

In [32]:
# 1g Can you compare these two models in terms of capturing the syntactic and the semantic relationship between words.

from metrics import compare_syntactic_relationships

syntactic_pairs = [
    ("book", "booked"),
    ("report", "reports"),
    ("close", "closed"),
    ("harmful", "harmless"),
]
# Filter pairs to include only those where both words exist in the vocabulary
valid_words = cbow_model.wv.index_to_key
valid_syntactic_pairs = [
    (word1, word2) for word1, word2 in syntactic_pairs if word1 in valid_words
]

# Comparing syntactic relationships
syntactic_results = compare_syntactic_relationships(
    valid_syntactic_pairs, skip_gram_model, cbow_model
)

# Print the results for syntactic relationships
print("Syntactic Relationships:")
for pair, result in syntactic_results.items():
    print(
        f"{pair[0]} - {pair[1]} | Skip-gram similarity: {result['skipgram']:.4f}, CBOW similarity: {result['cbow']:.4f}"
    )

Syntactic Relationships:
book - booked | Skip-gram similarity: 1.0000, CBOW similarity: 1.0000
report - reports | Skip-gram similarity: 1.0000, CBOW similarity: 1.0000
close - closed | Skip-gram similarity: 1.0000, CBOW similarity: 1.0000
harmful - harmless | Skip-gram similarity: 1.0000, CBOW similarity: 1.0000


------------------------------------------------------------------------------------------

# Section 2

In [1]:
from contextual import get_word_embedding, compute_cosine_similarity

# Sentences with polysemy for the word "lead"
sentence1 = "The lead actor delivered an incredible performance."
sentence2 = "Exposure to lead is harmful to health."

# Example usage: Get embeddings for "lead" in two different contexts and compute similarity
embedding1, embedding2 = get_word_embedding(sentence1, sentence2)

# Compute cosine similarity between the two embeddings
similarity = compute_cosine_similarity(embedding1, embedding2)
print(f"Cosine similarity between the two 'lead' embeddings: {similarity:.4f}")

Cosine similarity between the two 'lead' embeddings: 0.2524


------------------------------------------------------------------------------------------

# Section 3

# Metrics for Evaluating Word Embeddings

### 1. **Cosine Similarity**
   - **Why**: Measures the similarity between word vectors, ideal for evaluating both syntactic and semantic similarity.
   - **How to Collect**: Compare word pairs (e.g., **"man"** vs **"woman"**) using cosine similarity.

### 2. **Word Analogies Accuracy**
   - **Why**: Evaluates how well embeddings capture syntactic relationships (e.g., verb tense, singular/plural).
   - **How to Collect**: Use analogy datasets (e.g., **man:woman::king:queen**) and measure accuracy in solving analogies.

### 3. **Word Similarity Tasks (Spearman’s Rank Correlation)**
   - **Why**: Measures the correlation between model-generated and human-annotated similarity scores, useful for semantic evaluation.
   - **How to Collect**: Use word similarity datasets (e.g., **WordSim-353**, **SimLex-999**) and calculate Spearman’s rank correlation.

### 4. **Clustering Metrics (Silhouette Score, ARI)**
   - **Why**: Assesses how well words with similar meanings are grouped, useful for evaluating semantic relationships.
   - **How to Collect**: Cluster word embeddings and calculate clustering quality using Silhouette Score or ARI.

### 5. **Out-of-Vocabulary (OOV) Generalization**
   - **Why**: Tests how well embeddings generalize to unseen words, important for semantic tasks.
   - **How to Collect**: Evaluate embedding performance on unseen words in related contexts, measuring cosine similarity or prediction accuracy.

### 6. **Downstream Task Performance (Accuracy, F1-Score)**
   - **Why**: Measures the utility of embeddings in real-world tasks, providing an extrinsic evaluation of their effectiveness.
   - **How to Collect**: Use embeddings in tasks like sentiment analysis, NER, or text classification and report task-specific metrics (e.g., accuracy, F1-score).


------------------------------------------------------------------------------------------
# Section 4

In [1]:
from custom_w2v import evaluate_word_pairs
from word_2_vec import read_sentences, train_skipgram_model

custom_corpus_file_path = "data/custom_corpus.txt"
custom_corpus = read_sentences(custom_corpus_file_path)

custom_skip_gram_model = train_skipgram_model(
    custom_corpus, vector_size=128, window=3, min_count=0
)

# Word pairs for evaluation
word_pairs = [("king", "queen"), ("man", "woman"), ("apple", "orange")]

evaluate_word_pairs(custom_skip_gram_model, word_pairs)

Cosine similarity between 'king' and 'queen': 0.9516
Cosine similarity between 'man' and 'woman': 0.9753
Cosine similarity between 'apple' and 'orange': 0.9825


# Cosine Similarity Results

The following results represent the **cosine similarity** between different word pairs. Cosine similarity measures the cosine of the angle between two vectors, indicating how similar the words are in the **embedding space**. A value closer to **1** indicates that the words are very similar in meaning, while a value closer to **0** suggests they are less similar.

### Results:

- **Cosine similarity between 'king' and 'queen': 0.9516**
   - The high similarity (0.9516) between **"king"** and **"queen"** reflects the strong **semantic relationship** between the two words. These are closely related in terms of royalty and gender, and Word2Vec captures this semantic proximity well.

- **Cosine similarity between 'man' and 'woman': 0.9753**
   - A similarity of **0.9753** indicates an extremely close relationship between **"man"** and **"woman"**. This pair represents **gendered counterparts**, and the high similarity score shows how well the model captures this relationship in the word embeddings.

- **Cosine similarity between 'apple' and 'orange': 0.9825**
   - The **0.9825** similarity between **"apple"** and **"orange"** demonstrates the strong **semantic similarity** between these two fruits. Word2Vec captures their similarity as they are both common fruits, indicating that the model successfully recognizes their related meanings.

### Conclusion:
- The cosine similarity values for the word pairs **king:queen**, **man:woman**, and **apple:orange** are all quite high, suggesting that the **Word2Vec** model has effectively learned the semantic relationships between gendered terms and similar objects (fruits) in the embedding space.


------------------------------------------------------------------------------------------
# Section 5